portableService: tooling to create portable service images

gdamjan opened the pull request:
see https://systemd.io/PORTABLE_SERVICES/ about the definition of
portable services. this tooling is analogous to the `pkgs.dockerTools.buildImage`
tooling and is currently called `pkgs.portableService`.

systemd (since version 239) supports a concept of “Portable Services”.
“Portable Services” are a delivery method for system services that uses
two specific features of container management:

* Applications are bundled. I.e. multiple services, their binaries and
  all their dependencies are packaged in an image, and are run directly from it.
* Stricter default security policies, i.e. sandboxing of applications.

The primary tool for interacting with Portable Services is portablectl,
and they are managed by the systemd-portabled service.

This function will create a squashfs raw image in result/$name_$version.raw
that has the required files by the portable services spec, and all the dependencies
for the running program in the nix store.



###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

gdamjan commented:
I'm opening this PR for discussion firstly. I'll need help in writing some documentation (please refer me where to add it), and perhaps in creating some tests (again help wanted).

Some example usage can be seen here:
https://github.com/gdamjan/wallabag-service/blob/main/default.nix
and is based on my previous work, where I first implemented the creation of the portable images:
https://github.com/gdamjan/nextcloud-service
https://github.com/gdamjan/tt-rss-service

bor0 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,75 @@
+      cp ${os-release} $out/etc/os-release
+
+      # units **must** be copied to /etc/systemd/system/
+      ${lib.forEach units (u: "cp ${u} $out/etc/systemd/system/${u.name};")}
```
I guess?

```suggestion
      ${lib.forEach units (u: "cp ${u.name} $out/etc/systemd/system/${u.name};")}
```

gdamjan replied:
so (I wish there were type annotations in Nix 🙂) `units` is expected to be a list of derivations. The derivation has a name attribute, but when used as ${u} in a shell/string context, it will evaluate to the actual path in the nix store.

```
nix-repl> wallabag-service = pkgs.substituteAll { name = "wallabag.service"; src = ./fajl; world = "test"; }

nix-repl> wallabag-service
«derivation /nix/store/4wrl4jsl5li29sk9bry96f7s14imcg25-wallabag.service.drv»

nix-repl> wallabag-service.name
"wallabag.service"

nix-repl> "${wallabag-service}" 
"/nix/store/0674kr7idpjqb89maja8f6vlgzfxcjpv-wallabag.service"

nix-repl> "${wallabag-service.name}"
"wallabag.service"
```

bor0 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,75 @@
+  https://systemd.io/PORTABLE_SERVICES/
+
+  Example:
+  …
```
Was there supposed to be an example here?

gdamjan replied:
yes, I'll add a example once all is done

bor0 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,75 @@
+      # units **must** be copied to /etc/systemd/system/
+      ${lib.forEach units (u: "cp ${u} $out/etc/systemd/system/${u.name};")}
+
+      #for symlink in {symlinks}; do
+      #  cp $ $out/etc/systemd/system/
+      #done
```
Is this needed?

gdamjan replied:
that needs to be implemented, I was looking for a more elegant solution than the one in `nixos/lib/make-system-tarball.nix` and friends, but I guess I'll just copy/paste it now.

bor0 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,75 @@
+    done
+
+    mksquashfs nix ${rootfs}/* $out \
+      -quiet -noappend \
```
Does `-quiet` suppress errors as well?

gdamjan replied:
mksquashfs is unnecessary verbose by default. this quiets it down. it shouldn't hide errors.

alas, good point, I might need to add `-exit-on-error` just in case (by default it skips files that it can't add to the archive).

gdamjan commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+, squash-block-size ? "1M"
+}:
+
+let
```
I guess I should add an assert for the naming of the units (the must be prefixed with the name of the portable service)
```suggestion
let
  assertUnitNames = lib.assertMsg (lib.all (u: lib.hasPrefix name u.name) units) ''
      Unit names must be prefixed with the service name
  '';
```

gdamjan commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+  imageFileName = "${name}_${version}.raw";
+  imageDrvName = "${name}-img";
+
+in
```
```suggestion
in

assert assertUnitNames
```

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+
+  Example:
+  pkgs.portableService {
+      name = "demo";
+      units = [ demo-service demo-socket ];
```
```suggestion
    name = "demo";
    units = [ demo-service demo-socket ];
```

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+{
+  # The name and version of the portable service. The resulting image will be
+  # created in result/$name_$version.raw
+  name
```
```suggestion
  pname
```
maybe just follow mkDerivation pname+version.

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+
+    in
+    stdenv.mkDerivation {
+      name = "root-fs-scaffold";
```
```suggestion
      pname = "root-fs-scaffold";
      inherit version;
```

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+        ${lib.concatMapStringsSep "\n" (u: "cp ${u} $out/etc/systemd/system/${u.name};") units}
+
+        # Add symlinks to the top-level store objects.
+        symlinkTargets=($symlinkTargets)
+        symlinkNames=($symlinkNames)
+        for ((n = 0; n < ''${#symlinkTargets[*]}; n++)); do
+          target=''${symlinkTargets[$n]}
+          symlink=''${symlinkNames[$n]}
+          if test "$symlink" != "none"; then
+              mkdir -p $(dirname "$out/$symlink")
+              ln -sT "$target" "$out/$symlink"
+          fi
+        done
```
I would recommend to use nix code to generate this rather than relying on respliting bash variables.

gdamjan replied:
indeed, I copied that from some other .nix files. Let me rethink that

gdamjan replied:
something like this 340257ba5ec3094eff062c0439f136483e97eff7 ?

SuperSandro2000 replied:
I just took a brief look but that already looks good

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+
+  # the '.raw' suffix is mandatory by the portable service spec
+  imageFileName = "${name}_${version}.raw";
+  imageDrvName = "${name}-img";
+
+in
+
+stdenv.mkDerivation {
+  name = imageDrvName;
```
```suggestion
in

stdenv.mkDerivation {
  pname = "${name}-img";
  inherit version:

```

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+        done
+      '';
+    };
+
+  # the '.raw' suffix is mandatory by the portable service spec
+  imageFileName = "${name}_${version}.raw";
```
```suggestion
```

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+    done
+
+    mkdir -p $out
+    mksquashfs nix ${rootFsScaffold}/* $out/${imageFileName} \
```
```suggestion
    # the '.raw' suffix is mandatory by the portable service spec
    mksquashfs nix ${rootFsScaffold}/* $out/"${pname}_${version}.raw" \
```

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+  nativeBuildInputs = [ squashfsTools ];
+  closureInfo = pkgs.closureInfo { rootPaths = [ rootFsScaffold ] ++ contents; };
+
+  buildCommand = ''
+
+    mkdir -p nix/store
```
```suggestion
  buildCommand = ''
    mkdir -p nix/store
```

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,118 @@
+  # The name and version of the portable service. The resulting image will be
+  # created in result/$name_$version.raw
+  name
+, version ? "dev"
```
I am unsure about this and don't really want to have the default here while I can understand why you added it.

gdamjan replied:
ok. removed that. `version` is mandatory now.

gdamjan commented:
should I run some source code formatter on the file? and which one?

bor0 commented:
Code wise this looks good. Thanks for addressing my comments.

Note that I haven't tested this at all. If you want me to do that as well it'd be great to add some detailed testing instructions to the description, as I'm a complete beginner to this repo :)

gdamjan commented:
Testing with docker (or podman):

```
git clone https://gist.github.com/gdamjan/9f3e4a4016459d3e12f576ec0813c785.git
git clone -b make-portable https://github.com/gdamjan/nixpkgs.git

# run an interactive docker (or podman)
docker run -ti --rm \
  -v ./9f3e4a4016459d3e12f576ec0813c785:/work \
  -v ./nixpkgs:/nixpkgs \
  -e NIX_PATH=nixpkgs=/nixpkgs \
  nixos/nix:latest

# inside the container:
nix-build /work/default.nix
ls result/
sha256sum result/demo_1.0.raw
```

you should get `4fe5cb3ceabf8eae59b5cccda64ea380fc3ce1406bd1584551fff418246c5fff  result/demo_1.0.raw`

bor0 commented:
```
bash-5.1# nix-build /work/default.nix
...

/nix/store/ycpsvd5fi4yryz904m2d6hjs9zmvbbh9-demo-img-1.0
bash-5.1# ls result/
demo_1.0.raw
bash-5.1# sha256sum result/demo_1.0.raw
4fe5cb3ceabf8eae59b5cccda64ea380fc3ce1406bd1584551fff418246c5fff  result/demo_1.0.raw
```

SuperSandro2000 commented:
> should I run some source code formatter on the file? and which one?

https://github.com/nix-community/nixpkgs-fmt

Also please squash the commits from review into the init commit.

gdamjan commented:
> > should I run some source code formatter on the file? and which one?
> 
> https://github.com/nix-community/nixpkgs-fmt

I did, the result is a bit strange at the fist comment, otherwise looks good
https://github.com/NixOS/nixpkgs/pull/161278/commits/31427aa9b25988ab1698f4ccddde22d3f21df7af#diff-e9c0525951f869d69d7be9b1d0b854cf510b38e2effb8007a4518b6dfe1253ffR8-R9

> Also please squash the commits from review into the init commit.

I will, just check the above please :)

gdamjan committed:
portableService: tooling to create portable service images

see https://systemd.io/PORTABLE_SERVICES/ about the definition of
portable services. this tooling is analogous to the `pkgs.dockerTools.buildImage`
tooling and is called `pkgs.portableService`.

systemd (since version 239) supports a concept of “Portable Services”.
“Portable Services” are a delivery method for system services that uses
two specific features of container management:

* Applications are bundled. I.e. multiple services, their binaries and all
  their dependencies are packaged in an image, and are run directly from it.
* Stricter default security policies, i.e. sandboxing of applications.

The primary tool for interacting with Portable Services is portablectl,
and they are managed by the systemd-portabled service.

This function will create a squashfs raw image in `result/$pname_$version.raw`
that has the required files by the portable services spec, and all the
dependencies for the running program in the nix store.

gdamjan commented:
squashed and manually reverted the nixpkgs-fmt from the introduction comment

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,111 @@
+}:
+
+let
+  assertUnitNames = lib.assertMsg (lib.all (u: lib.hasPrefix pname u.name) units) ''
+    Unit names must be prefixed with the service name
+  '';
+  filterNull = lib.filterAttrs (_: v: v != null);
```
```suggestion
  filterNull = lib.filterAttrs (_: v: v != null);
```

SuperSandro2000 commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,111 @@
+    };
+in
+
+assert assertUnitNames;
```
```suggestion
assert lib.assertMsg (lib.all (u: lib.hasPrefix pname u.name) units) "Unit names must be prefixed with the service name";

```

SuperSandro2000 reviewed (approved):
otherwise LGTM

gdamjan commented:
> otherwise LGTM

thanks. fixed, rebased, squashed!

RaitoBezarius commented on pkgs/build-support/portable-service/default.nix:
```diff
@@ -0,0 +1,109 @@
+        cp ${os-release} $out/etc/os-release
+      ''
+      # units **must** be copied to /etc/systemd/system/
+      + (lib.concatMapStringsSep "\n" (u: "cp ${u} $out/etc/systemd/system/${u.name};") units)
```
why not inlinining `${lib.concatMapStringsSep ...}` and not doing string concatenation?

RaitoBezarius replied:
inlining the Nix code is not doing bash stuff right?

aanderse replied:
I think @RaitoBezarius means this:

```
      buildCommand = ''
        # scaffold a file system layout
        mkdir -p $out/etc/systemd/system $out/proc $out/sys $out/dev $out/run \
                 $out/tmp $out/var/tmp $out/var/lib $out/var/cache $out/var/log
        # empty files to mount over with host's version
        touch $out/etc/resolv.conf $out/etc/machine-id
        # required for portable services
        cp ${os-release} $out/etc/os-release

      # units **must** be copied to /etc/systemd/system/
      ${lib.concatMapStringsSep "\n" (u: "cp ${u} $out/etc/systemd/system/${u.name};") units}
```

I don't think it really matters one way or the other.

aanderse commented:
This is **super** cool for certain use cases. Super cool. What is it going to take to get this merged?

gdamjan commented:
> This is **super** cool for certain use cases. Super cool. What is it going to take to get this merged?

is something else expected from me? I'd be happy to improve the PR.

phaer commented:
@gdamjan Portable Services seem to support not only images, but directories as well. Have you considered adding support for the latter? It might be overkill for this PR and could be a follow up, just not sure about the functions interface yet.

Supporting directories might make it easier to build portable services which just mount the hosts nix store. This would make them less portable, as that obviously would require NixOS on the host, but could be useful to benefit from the sandboxing.

aanderse commented:
The only thing I could think to add is some documentation. I think the implementation of this is good 👍

I would be happy to see this merged and iterate on documentation after the fact.

gdamjan commented:
> @gdamjan Portable Services seem to support not only images, but directories as well. Have you considered adding support for the latter? It might be overkill for this PR and could be a follow up, just not sure about the functions interface yet.

let's take that to the matrix channels (-dev, -systemd or nixos)

> The only thing I could think to add is some documentation. I think the implementation of this is good +1
> I would be happy to see this merged and iterate on documentation after the fact.

thanks, I'll work on the documentation in another PR.
Can you share a pointer how should it look, and where to add it please?
